### ISSUE TITLE
[java] FN in ImmutableField for field with a default + constructor assignment

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#4869]
Under PMD 7.0.0 there is no report for the code:

```java
public class NotSoImmutableField {
    private int foo = 2;

    NotSoImmutableField() {
    }

    NotSoImmutableField(final int foo) {
        this.foo = foo;
    }

    public int getFoo() {
        return foo;
    }
}
```

The field is never written more than once within a given constructor, nor depends on the previous value. Therefore, the code **can** effectively be rewritten to make `foo` immutable, as it's never written after initialization, for instance, the code could be rewritten as:

```java
public class NotSoImmutableField {
    private final int foo;

    NotSoImmutableField() {
        this(2);
    }

    NotSoImmutableField(final int foo) {
        this.foo = foo;
    }

    public int getFoo() {
        return foo;
    }
}
```

_Originally posted by @jsotuyod in https://github.com/pmd/pmd/issues/4046#issuecomment-2012592805_

**Comment by oowekyala** [comment]
I think this makes sense in theory, but it probably should be a new rule. I like that ImmutableField always recommends making things final if it actually can be done by just adding a final keyword, and there is no chance the compiler will reject the new code. The fix could in theory be done automatically. Recommending that the user adapts their initialization logic like you suggest is a different task for the programmer and requires more effort, also there are different ways to refactor that. The rule implementation would also likely be very different... So yeah I think it should be a new rule.
